MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+prefetch packs will appear and occasionally a very large prefetch pack covering
+everything more than 30 days old. This large prefetch pack appears on initial
+clone and sometimes again if a user does not prefetch for 30 days (this occurs 
+less often now that background prefetch exists).
+
+### Delete pack-files with no objects referenced by the multi-pack-index
+
+The multi-pack-index stores a single copy of every packed object. When an object
+appears in multiple pack-files covered by the multi-pack-index, we select the
+pack-file with the most-recent _modified time_ as reported by the filesystem.
+For example, if someone downloads a second large prefetch pack when prefetching
+for the first time in 31 days, then that large prefetch pack is referenced and
+all older prefetch packs are "stale" and not referenced by the multi-pack-index.
+
+We can extend the `git multi-pack-index` builtin to delete pack-files that have
+no objects referenced by the multi-pack-index. This should happen at the same
+time as we update the multi-pack-index since we don't want to have this list of
+pack-files in the multi-pack-index if they don't exist. We also don't want to
+remove them from the multi-pack-index while the pack-files still exist, because
+then a Git process will start referencing those pack-files in the `packed_git`
+linked list.
+
+### Repack the prefetch pack-files
+
+We can reduce the number of prefetch packs and improve file locality by
+combining prefetch packs into new "repacked" pack-files. Use `git show-index
+<object-dir>/pack/prefetch-<timestamp[i]>-<hash[i]>.idx` to list the objects
+in the prefetch packs and pipe that output to `git pack-objects -q
+<object-dir>/pack/repacked-<timestamp>`.
+
+We should refrain from packing the most-recent prefetch packs, as we don't want
+to get in the way of the prefetch machinery for determining the most-recent
+timestamp. We could, however, start repacking the prefetch packs from
+oldest-to-newest, stopping in the most-recent week, or when the prefetch packs
+are large enough to satisfy a batch. Some stopping conditions could include
+"pack-file size is above 1GB".
+
+Running the Jobs
+----------------
+
+We already have a post-fetch job that is automatically queued in the background.
+We should rename this job to be a "maintenance" job that could trigger several
+actions. The maintenance job should run every 15 minutes (as the post-fetch job
+currently does) but then it should check each "sub-job" for a timing condition
+to run again.
+
+For example:
+
+* The post-fetch job doesn't trigger if the most-recent prefetch pack timestamp
+  is within the last 70 minutes.
+* The loose object job doesn't trigger if the most-recent "loose" pack timestamp
+  is within the last 24 hours.
+* The prefetch repack job doesn't trigger if the most-recent "repacked" pack
+  timestamp is within the last 7 days or the most-recent "repacked" pack is
+  modified within the last 24 hours.

```

Comment:
How did you come up with these numbers for running?  Is there data to back up this being the most efficient or is it based on the prefetch packs?

Reply:
The loose timing comes from "24 hours should be enough for no Git processes to survive, and not too obtrusive to users". This could be flexible, depending on how often we think these loose objects can be filled.

The prefetch repack timing is based on looking at the shared object cache and seeing the size of the hourly and daily prefetch packs for the Windows repo.

* Looks like recent activity has been around 80MB a day. This amounts to less than 500 MB a week, so maybe it isn't enough to fill a 1GB batch. This does raise a good point: we should leave a timestamp file for "most-recent check for a batch" to avoid repeated checks for "do we have a large enough batch?".

* The 24 limit on the modified time is to allow us to catch up for really old repos with a long history of prefetch packs, but also because we want to delete pack-files on the next run.